modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts

```diff
@@ -0,0 +1,115 @@
+
+    static fromErc4626Token(erc4626Token: Erc4626PoolToken): BufferPool {
+        const mainToken = new BasePoolToken(erc4626Token.token, MAX_UINT256, 0);
+        const underlyingToken = new BasePoolToken(
+            new Token(
+                erc4626Token.token.chainId,
+                erc4626Token.underlyingTokenAddress as Address,
+                erc4626Token.token.decimals,
+            ),
+            MAX_UINT256,
+            1,
+        );
+        return new BufferPool(
+            erc4626Token.token.address,
+            erc4626Token.token.address,
+            erc4626Token.token.chainId,
+            erc4626Token.rate,
+            [mainToken, underlyingToken],
+        );
+    }
+
+    constructor(id: Hex, address: string, chainId: number, rate: bigint, tokens: BasePoolToken[]) {
+        this.chainId = chainId;
+        this.id = id;
+        this.address = address;
+        this.rate = rate;
+        this.tokens = tokens;
+        this.tokenMap = new Map(tokens.map((token) => [token.token.address, token]));
+    }
+
+    public getNormalizedLiquidity(tokenIn: Token, tokenOut: Token): bigint {
+        return MAX_UINT256;
+    }
+
+    public getLimitAmountSwap(tokenIn: Token, tokenOut: Token, swapKind: SwapKind): bigint {
+        return MAX_UINT256;
+    }
+
+    public swapGivenIn(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
+
+        const poolState = this.getPoolState();
+        const vault = new Vault();
+
+        const calculatedAmount = vault.swap(
+            {
+                amountRaw: swapAmount.amount,
+                tokenIn: tIn.token.address,
+                tokenOut: tOut.token.address,
+                swapKind: SwapKind.GivenIn,
+            },
+            poolState,
+        );
+        return TokenAmount.fromRawAmount(tOut.token, calculatedAmount);
+    }
+
+    public swapGivenOut(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
+
+        const poolState = this.getPoolState();
+        const vault = new Vault();
+
+        // swap
+        const calculatedAmount = vault.swap(
+            {
+                amountRaw: swapAmount.amount,
+                tokenIn: tIn.token.address,
+                tokenOut: tOut.token.address,
+                swapKind: SwapKind.GivenOut,
+            },
+            poolState,
+        );
+        return TokenAmount.fromRawAmount(tIn.token, calculatedAmount);
+    }
+
+    public getPoolState(): BufferState {
+        return {
+            poolType: 'Buffer',
+            poolAddress: this.address,
+            tokens: this.tokens.map((t) => t.token.address),
+            rate: this.rate,
+        };
+    }
+
+    // Helper methods
+
+    public getRequiredTokenPair(tokenIn: Token, tokenOut: Token): { tIn: BasePoolToken; tOut: BasePoolToken } {
```

Comment:
For some reason I find it hard to reason about whats happening here - could you explain and maybe refactor/rename to make it a bit clearer?

Reply:
This is actually doing a two things:
1. validating that the pool indeed contains tokenIn and tokenOut
2. getting pool tokens with specific attributes (e.g. weight for WeightedPool, rate for StablePool) and their state (balances, index, etc..) for tokenIn and tokenOut

I agree it could be more clear on its purpose.
One thing that I think might help is to add it to the BasePool interface to highlight that this is present on all pools.
Do you have any specific naming suggestions? I couldn't think of a better naming for it, but I'll add some comments explaining what it's doing.

Let me know what you think.

Reply:
Ahh right, got it, thanks! Comments will definitely help clarify.

Maybe something like `getBasePoolTokens` might make it clearer?

Looks like its just using the wrappedToken address so might also make sense to make the input params `0x${string}` types instead of `Token` and rename to something like wrappedTokenIn/Out? Can you remind me of what the `wrapped` param represents and what it would be for something like stataUSDC? Thanks!

Reply:
I think I'll go with `getPoolTokens` so it's more comprehensive for all PoolToken types.

About updating the inputs, I don't think there's much benefit from it. This method is used only internally on each pool, so I think it's simpler to stick with forwarding Token from external inputs (e.g. swapGivenIn, swapGivenOut, etc.). I can revert it back to private and remove it from the interface if you think it makes more sense.

`wrapped` is used only in the context of native assets and their wrapped versions (e.g. ETH/WETH). For all other tokens, `address` and `wrapped` are equal.

Reply:
Cool. Just one last question on this - doesn't look like any of the other pools are checking `wrapped` should they?

Reply:
Yes, you're right. We should be checking wrapped everywhere. Very few places should check for address directly. I'll fix that over here.

Reply:
Although, looking again at the code, it seems the API is handling this differently and replacing ETH with WETH inputs so this `wrapped` concept may not even make much sense anymore in the current scenario. I'll create a separate task to investigate this and possibly remove that from SDK/SOR Token type.